swift/Arrow/Sources/Arrow/ArrowCImporter.swift

```diff
@@ -0,0 +1,172 @@
+            ArrowCImporter.release(cSchema)
+            return .success(field)
+        case .failure(let err):
+            ArrowCImporter.release(cSchema)
+            return .failure(err)
+        }
+    }
+
+    public func importArray(
+        _ cArray: UnsafePointer<ArrowC.ArrowArray>,
+        arrowType: ArrowType,
+        isNullable: Bool = false
+    ) -> Result<ArrowArrayHolder, ArrowError> {
+        let arrowField = ArrowField("", type: arrowType, isNullable: isNullable)
+        return importArray(cArray, arrowField: arrowField)
+    }
+
+    public func importArray( // swiftlint:disable:this cyclomatic_complexity
+        _ cArrayPtr: UnsafePointer<ArrowC.ArrowArray>,
+        arrowField: ArrowField
+    ) -> Result<ArrowArrayHolder, ArrowError> {
+        let cArray = cArrayPtr.pointee
+        if cArray.null_count < 0 {
+            ArrowCImporter.release(cArrayPtr)
+            return .failure(.invalid("Uncomputed null count is not supported"))
+        } else if cArray.n_children > 0 {
+            ArrowCImporter.release(cArrayPtr)
+            return .failure(.invalid("Children currently not supported"))
+        } else if cArray.dictionary != nil {
+            ArrowCImporter.release(cArrayPtr)
+            return .failure(.invalid("Dictionary types currently not supported"))
+        } else if cArray.offset != 0 {
+            ArrowCImporter.release(cArrayPtr)
+            return .failure(.invalid("Offset of 0 is required but found offset: \(cArray.offset)"))
+        }
+
+        let arrowType = arrowField.type
+        let length = UInt(cArray.length)
+        let nullCount = UInt(cArray.null_count)
+        var arrowBuffers = [ArrowBuffer]()
+
+        if cArray.n_buffers > 0 {
+            if cArray.buffers == nil {
+                ArrowCImporter.release(cArrayPtr)
+                return .failure(.invalid("C array buffers is nil"))
+            }
+
+            switch arrowType.info {
+            case .variableInfo:
+                if cArray.n_buffers != 3 {
+                    ArrowCImporter.release(cArrayPtr)
+                    return .failure(
+                        .invalid("Variable buffer count expected 3 but found \(cArray.n_buffers)"))
+                }
+
+                appendToBuffer(cArray.buffers[0], arrowBuffers: &arrowBuffers, length: length)
+                appendToBuffer(cArray.buffers[1], arrowBuffers: &arrowBuffers, length: length)
+                appendToBuffer(cArray.buffers[2], arrowBuffers: &arrowBuffers, length: length)
+            default:
+                if cArray.n_buffers != 2 {
+                    ArrowCImporter.release(cArrayPtr)
+                    return .failure(.invalid("Expected buffer count 2 but found \(cArray.n_buffers)"))
+                }
+
+                appendToBuffer(cArray.buffers[0], arrowBuffers: &arrowBuffers, length: length)
+                appendToBuffer(cArray.buffers[1], arrowBuffers: &arrowBuffers, length: length)
```

Comment:
Same here, if we actually care about the buffers' bytelengths:
* for buffer 0 (null bitmap): same calculation as above
* for buffer 1 (primitive values): `(length + 7) / 8` if type is boolean, otherwise `length * sizeof(primitive type)`